plugins/org.eclipse.n4js.ts.model/model/Types.xcore

```diff
@@ -198,25 +219,52 @@ class TModule extends AbstractNamespace, SyntaxRelatedTElement, TAnnotableElemen
 	contains transient Type[] temporaryTypes
 
 	/**
-	 * Returns this module's module specifier as it would appear, for example, in the string literal at the end of an
-	 * import statement.
-	 * <p>
-	 * Since we changed the delimiter for our internal qualified names from '.' to '/', this simply returns the same
-	 * value as {@link #getQualifiedName()}. However, this getter is retained for the time being to let client code
-	 * differentiate between internal use (qualified name, e.g. in the Xtext index) and Javascript context (module
-	 * specifier, e.g. in import statements).
+	 * Convenience method, returns module this element is contained in, or
+	 * null if it is not contained in a module.
 	 */
-	derived String moduleSpecifier get {
-		return qualifiedName
+	op TModule getContainingModule() {
+		return this;
 	}
-	
+	/**
+	 * Convenience method, returns the {@link TModule} this element is contained in, or
+	 * null if it is not contained in a TModule. Does not consider intermediate {@link TDeclaredModule}s!
+	 */
+	op TModule getContainingRootModule() {
+		return this;
+	}
+}
+
+/*
+ * This type of module is created only from .d.ts files and represents a module created from
+ * a module declaration such as
+ * <pre>
+ * // .d.ts
+ * declare module "my/declared/module" { ... }
+ * </pre>
+ * Note that module declarations of the form
+ * <pre>
+ * // .d.ts
+ * declare module myLegacyModule { ... }
+ * </pre>
+ * are actually an old syntax for namespaces and will thus be represented as a {@link TNamespace},
+ * not as a {@code TDeclaredModule}.
+ */
+class TDeclaredModule extends AbstractModule {
```

Comment:
Adding to my comment above: I think it is a drawback that the grammar of N4JS is modified due to a pure d.ts feature. My assumption is, that his feature is not part of our roadmap for N4JS - what do you think? In other cases like the namespaces, it seems more likely that we actually want that as a N4JS feature.

Reply:
Yes, declared modules will most likely not be made available in N4JS[D].
(other than that, see response above)